WearTilesKotlin/app/src/main/java/com/example/wear/tiles/messaging/MessagingTileService.kt

```diff
@@ -35,77 +35,87 @@ import kotlinx.coroutines.flow.map
 import kotlinx.coroutines.flow.stateIn
 
 /**
- * Creates a Messaging Tile, showing your favorite contacts and a button to search other contacts.
- * This is a demo tile only, so the buttons don't actually work.
+ * Creates a Messaging tile, showing up to 4 contacts, an icon button and compact chip.
  *
- * The main function, [onTileRequest], is triggered when the system calls for a tile and implements
- * ListenableFuture which allows the Tile to be returned asynchronously.
+ * It extends [CoroutinesTileService], a Coroutine-friendly wrapper around
+ * [androidx.wear.tiles.TileService], and implements [tileRequest] and [resourcesRequest].
  *
- * Resources are provided with the [onResourcesRequest] method, which is triggered when the tile
- * uses an Image.
+ * The main function, [tileRequest], is triggered when the system calls for a tile. Resources are
+ * provided with the [resourcesRequest] method, which is triggered when the tile uses an Image.
  */
 class MessagingTileService : CoroutinesTileService() {
-    private lateinit var tileStateFlow: StateFlow<MessagingTileState?>
-    private lateinit var renderer: MessagingTileRenderer
     private lateinit var repo: MessagingRepo
     private lateinit var imageLoader: ImageLoader
+    private lateinit var renderer: MessagingTileRenderer
+    private lateinit var tileStateFlow: StateFlow<MessagingTileState?>
 
     override fun onCreate() {
         super.onCreate()
 
-        val appContainer = (application as TilesApplication).appContainer
-        renderer = appContainer.renderer
-        repo = appContainer.repo
-        imageLoader = ImageLoader(this)
+        repo = MessagingRepo(this)
+        imageLoader = Coil.imageLoader(this)
+        renderer = MessagingTileRenderer(this)
 
         tileStateFlow = repo.getFavoriteContacts()
-            .map {
-                buildState(it)
-            }
+            .map { contacts -> MessagingTileState(contacts) }
             .stateIn(
                 lifecycleScope,
                 started = SharingStarted.WhileSubscribed(5000),
                 initialValue = null
             )
     }
 
-    private fun buildState(contacts: List<Contact>): MessagingTileState {
-        return MessagingTileState(contacts)
-    }
-
+    /**
+     * Read the latest data, delegating to the [MessagingTileRenderer] which creates a layout to
```

Comment:
Is it useful to do a brief explanation of StateFlow and why it's used here?

Reply:
I would leave it for now - I think that it will make more sense if the data store is externally updated (e.g. from a configuration overlay)? At which point we could spend more time leaning into this.